File: lua/math/quick_pow.lua
local function quick_pow(x, n)
    x = (n > 0) and x or 1 / x
    n = (n > 0) and n or -n

    local ans = 1
    while n ~= 0 do
        if n & 1 == 1 then
            ans = ans * x
        end
        x = x * x
        n = n >> 1
    end
    return ans
end

print(quick_pow(2.1, 9))
print(quick_pow(2.1, -9))

File: zig/math/quick_pow.zig
const std = @import("std");

fn quick_pow(x: f32, n: i32) f32 {
    var x1: f32 = x;
    var n1: i32 = n;
    if (n < 0) {
        x1 = 1 / x;
        n1 = -n;
    }
    var ans: f32 = 1;
    while (n1 != 0) {
        if ((n1 & 1) == 1) {
            ans *= x1;
        }
        x1 *= x1;
        n1 >>= 1;
    }
    return ans;
}

pub fn main() void {
    std.debug.print("2^4: {}\n", .{quick_pow(2, 4)});
    std.debug.print("2^-4: {}\n", .{quick_pow(2, -4)});
    std.debug.print("2.1^6: {}\n", .{quick_pow(2.1, 6)});
    std.debug.print("2.1^-6: {}\n", .{quick_pow(2.1, -6)});
    std.debug.print("2.1^0: {}\n", .{quick_pow(2.1, 0)});
}
